nf_core/download.py

```diff
@@ -160,10 +161,53 @@ def download_singularity_image(self, container):
         """ Download singularity images for workflow """
 
         out_name = '{}.simg'.format(container.replace('nfcore', 'nf-core').replace('/','-').replace(':', '-'))
-        out_dir = os.path.abspath(os.path.join(self.outdir, 'singularity-images'))
+        out_path = os.path.abspath(os.path.join(self.outdir, 'singularity-images', out_name))
         address = 'docker://{}'.format(container.replace('docker://', ''))
-        singularity_command = ["singularity", "pull", "--name", os.path.join(out_dir, out_name), address]
-
+        shub_api_url = 'https://www.singularity-hub.org/api/container/{}'.format(container.replace('nfcore', 'nf-core').replace('docker://', ''))
+        shub_api_url = 'https://www.singularity-hub.org/api/container/ewels/nf-core-methylseq:latest'
+        singularity_command = ["singularity", "pull", "--name", out_path, address]
+
+        # Try to download the singularity image from singularity-hub first
+        logging.debug("Checking shub API: {}".format(shub_api_url))
+        response = requests.get(shub_api_url, timeout=10)
+        if response.status_code == 200:
+            shub_response = response.json()
+            logging.info("Downloading singularity image from singularity-hub.org")
+
+            # Strip the ?generation= key in URL, as this breaks the download
+
+            #
+            # TODO: May not need to do this?
+            #
+            dl_url = re.sub(r"\?generation=\d+&", '?', shub_response['image'])
+
+            # Stream the download as it's going to be large
+            logging.debug("Starting download: {}".format(dl_url))
+
+            #
+            # TODO: This next line hangs! WHHYYYYYY?Y???????
+            #
+            dl_request = requests.get(dl_url, stream=True)
+            if dl_request.status_code == 200:
+                logging.debug("Response code 200. Streaming download to disk.")
+                with open(out_path, 'wb') as f:
+                    for chunk in dl_request.iter_content(chunk_size=1024):
+                        if chunk:
+                            f.write(chunk)
```

Comment:
Please do `f.flush()` too :)

Reply:
This was intentional, due to [this stack overflow comment](https://stackoverflow.com/questions/16694907/how-to-download-large-file-in-python-with-requests-py#comment53495938_16696317)